pmlc/dialect/stripe/autostencil_pass.cc

```diff
@@ -0,0 +1,312 @@
+          if (arg0 != arg1) {
+            conflict[arg0].insert(arg1);
+            conflict[arg1].insert(arg0);
+          }
+        }
+      }
+    }
+  }
+  return used_idxs;
+}
+
+void AutoStencil::Transform() {
+  std::vector<std::pair<StringRef, size_t>> idxs = getAllIndex(cur_op);
+  llvm::SmallVector<int64_t, indexLimit> inner_sizes;
+  for (auto& idx : idxs) {
+    auto it = best_tiles.find(idx.first);
+    inner_sizes.push_back(it == best_tiles.end() ? 1 : it->second);
+  }
+  // The first tiling: split outer&middle and inner
+  Tile(cur_op, inner_sizes); 
+  ParallelForOp inner = mlir::dyn_cast<ParallelForOp>(cur_op.inner().front().front());
+
+  // Set ParallelForOp tags
+  setOpAttrUnit(cur_op, cur_op.getBodyBuilder(), "mac");
+  setOpAttrUnit(inner, inner.getBodyBuilder(), "mac_inner");
+  setOpAttrUnit(inner, inner.getBodyBuilder(), "xsmm");
+
+  // Set RefineOp tags
+  setOpAttrUnit(tensors[0]->getDefiningOp(), inner.getBodyBuilder(), "A");
+  setOpAttrUnit(tensors[1]->getDefiningOp(), inner.getBodyBuilder(), "B");
+  setOpAttrUnit(tensors[2]->getDefiningOp(), inner.getBodyBuilder(), "C");
+
+  // Set index tags
+  StringRef m_idx = idxName(inner_idxs[0]);
+  setIdxAttrUnit(inner, m_idx, "stencil");
+  setIdxAttrUnit(inner, m_idx, "stencil_m");
+  StringRef n_idx = idxName(inner_idxs[1]);
+  setIdxAttrUnit(inner, n_idx, "stencil");
+  setIdxAttrUnit(inner, n_idx, "stencil_n");
+  StringRef k_idx = idxName(inner_idxs[2]);
+  setIdxAttrUnit(inner, k_idx, "stencil");
+  setIdxAttrUnit(inner, k_idx, "stencil_k");
+}
+
+void AutoStencil::Stencil(ParallelForOp op) {
+  // Initialization
+  tensors.clear();
+  inner_idxs.clear();
+  best_perf = -1;
+
+  // Make a builder to write to just before the terminator
+  Block* obody = op.getBody();
+
+  // Collect all load ops to summarize the input tensors
+  obody->walk([&](LoadOp op) {
+    tensors.push_back(op.from());
+  });
+  if (tensors.size() != 2) {
+    return;
+  }
+  // Collect all store ops to summarize the output tensors
+  obody->walk([&](AggregateOp op) {
+    tensors.push_back(op.into());
+  });
+  if (tensors.size() != 3) {
+    return;
+  }
+
+  // The last tensor is the output.
+  out_idxs = RefUsedIdxs(tensors[numIndex - 1], true);
+  acc_idxs.clear();
+  for (size_t i = 0; i < numIndex - 1; ++i) {
+    BlockArgumentSet used_idxs = RefUsedIdxs(tensors[i], true);
+    for (auto idx : used_idxs) {
+      if (out_idxs.find(idx) == out_idxs.end()) {
+        acc_idxs.insert(idx);
+      }
+    }
+  }
+  all_idxs = acc_idxs;
+  all_idxs.insert(out_idxs.begin(), out_idxs.end());
+
+  cur_op = op;
+  // Search inner index and their tiles
+  SearchIndex(0);
+
+  // Transform
+  Transform();
+}
+
+void AutoStencilPass::runOnFunction() {
+  auto reqs = vertexai::tile::stripe::FromProto(options.reqs());
+  mlir::FuncOp f = getFunction();
+  AutoStencil as(options.processors());
+  f.walk([&reqs, &as] (ParallelForOp op) {
+    if (hasAttrs(op.getOperation(), reqs)) {
```

Comment:
I think using tag based approach to find blocks to modify is good right now for transition, however long term it might be good to just look for blocks that have the right form.